app/services/blog_service.rb

```diff
@@ -0,0 +1,101 @@
+require 'net/http'
+
+require 'uri'
+require 'active_support/time'
+
+class BlogService
+  CACHE_EXPIRY_PERIOD = 1.hour
+  CACHE_ERROR_PERIOD = 2.hours
+  RETRY_PERIOD = 1.second
+
+  attr_reader :key, :url, :retries, :force_fetch
+
+  def initialize(url: 'https://blog.energysparks.uk/feed/', retries: 0, force_fetch: false)
+    @url = url
+    @key = "blog:#{url}"
+    @retries = retries
+    @force_fetch = force_fetch
+  end
+
+  def cached_feed
+    Rails.cache.fetch(key)
+  end
+
+  def cache_out_of_date?(period)
+    !cached_feed || cached_feed&.dig(:timestamp) && cached_feed[:timestamp] < period.ago
+  end
+
+  def cache_expired?
+    cache_out_of_date?(CACHE_EXPIRY_PERIOD)
+  end
+
+  def caching_on?
+    !!Rails.application.config.action_controller.perform_caching
+  end
+
+  def fetch_feed?
+    cache_expired? || force_fetch
+  end
+
+  def raise_error?
+    caching_on? && cache_out_of_date?(CACHE_ERROR_PERIOD)
+  end
+
+  def tries
+    retries + 1
+  end
+
+  def items
+    if fetch_feed? && (feed = fetch_feed)
+      Rails.cache.write(key, feed)
+      return feed&.dig(:items)
+    elsif raise_error?
+      Rollbar.error("Blog cache for: #{url} is over #{CACHE_ERROR_PERIOD.seconds.in_hours} hours out of date")
+    end
+    cached_feed&.dig(:items)
+  end
+
+  # Can be run from cron if we don't want to tie up the front-page request
+  def update_cache!
+    Rails.cache.write(key, fetch_feed) if fetch_feed?
+  end
+
+  # Can be run in cron to decouple from this tie-ing up the front page request
+  def fetch_feed
```

Comment:
I think you can use the Faraday Retry middleware to do a fix number of retries with a back-off. We're using that elsewhere. https://github.com/lostisland/faraday-retry

Reply:
Thanks - I should have thought to check existing ways of doing this in the codebase. Makes things easier, thanks.